### PR TITLE
[None][perf] Fuse Kimi K3 KDA projections

### PR DESCRIPTION
**Pull request by jiaganc**
## Dev Engineer Review

- Fuses Kimi K3 KDA projections for prefill, decode, and MTP verification.
- Adds bounded auxiliary-stream overlap for workloads up to 128 rows.
- Preserves sequential fallbacks and the FP8 QKVG weight-read path.
- Extends `KimiKDARuntime.__init__` with optional `aux_stream`.
- Review should verify stream synchronization, FP8 compatibility, fallback behavior, and API consistency.

## QA Engineer Review

- Added `test_kda_fused_prefill_matches_separate_projections`.
- Added `test_kda_qkvg_multistream_decode_matches_separate_projections`.
- Extended two-round fused verification parity coverage.
- Tests compare projections, replay outputs, cache updates, and committed SSM state.
- CI coverage is registered in `tests/integration/test_lists/test-db/l0_b200.yml` and `tests/integration/test_lists/test-db/l0_gb300_multi_gpus.yml`.
- Targeted B200 validation passed with 7 passed, 0 failed, and 0 skipped.
- Verdict: sufficient.

## Description

Optimize Kimi K3 KDA projections in the PyTorch backend following the SGLang projection layout:

- replace the decode-only `[q|k|v|g|f_a|b]` projection with an alignment-friendly fused QKVG projection and a padded fused `[f_a|b]` projection;
- use the same fused projections during prefill and reuse QKVG's output-gate result;
- use the fused QKVG and `[f_a|b]` projections in both MTP verification implementations, including output-gate reuse;
- overlap `[f_a|b] -> f_b` with QKVG on the model auxiliary CUDA stream for decode and MTP workloads containing at most 128 rows;
- preserve sequential fallbacks and the existing FP8 QKVG weight-read path.

Keeping the small `f_a` and `b` tails out of the wide QKVG output avoids degrading GEMM kernel selection while reducing projection launches in prefill, decode, and MTP verification.

## Performance

+2% throughput perf gain

## Test Coverage

- Added public-forward parity coverage for fused prefill versus separate projections.
- Added public-forward parity coverage for multi-stream decode versus separate projections, including convolution and recurrent-cache updates.
- Extended the random-weight two-round MTP test to compare fused and sequential verification with mixed accepted-token counts and committed SSM state.
- Initialized the older sequential parity test's load-time weight buffers so it cannot compare NaN outputs from uninitialized tensors.
- Passed all pre-commit hooks and `git diff --check`.
- Built the native targets for SM100 on an NVIDIA B200.
- Passed `test_kimi_kda_verify_parity.py` and `test_kimi_kda_fused_verify_parity.py` on the B200: `7 passed, 0 failed, 0 skipped`.
- The GPU tests construct deterministic random weights and do not load Kimi K3 checkpoint weights.

## PR Checklist

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py [hunk 1]
```diff
@@ -1575,13 +1683,23 @@ def _forward_verify_fused(
...
+            q_proj, k_proj, v_proj, forget_gate, beta_proj, onorm_g = projections
+        x_q = q_proj.view(1, T_total, H, K)
+        x_k = k_proj.view(1, T_total, H, K)
+        x_v = v_proj.view(1, T_total, H, mixer.head_dim)
         # Raw gate / beta: the kernel applies dt_bias, A_log, the
         # lower-bound sigmoid gate, and the beta sigmoid itself.
-        g = mixer.f_b_proj(mixer.f_a_proj(x)).view(1, T_total, H, K)
-        beta = mixer.b_proj(x).view(1, T_total, H)
+        g = forget_gate.view(1, T_total, H, K)
+        beta = beta_proj.contiguous().view(1, T_total, H)
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_kimi_linear.py:1709]: Nit: should `_project_verify_inputs` handle the contiguity? Feels a bit weird for the caller(s) to do it.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py [hunk 2]
```diff
@@ -1236,9 +1246,27 @@ def _forward_prefill(
...
-        q_proj_states = mixer.q_proj(x)
-        k_proj_states = mixer.k_proj(x)
-        v_proj_states = mixer.v_proj(x)
+        onorm_g = None
+        if self._qkvg_proj_weight is not None:
+            qkvg = torch.nn.functional.linear(x, self._qkvg_proj_weight)
+            q_proj_states, k_proj_states, v_proj_states = qkvg[..., : 3 * d].split(d, dim=-1)
+            onorm_g = qkvg[..., 3 * d : 4 * d]
+        else:
+            fused_qkvg = getattr(mixer, "qkvg_proj", None)
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_kimi_linear.py:1262]: This FP8 branch (and the matching one in `_project_verify_inputs`) is new behavior with no test coverage: with the FP8 weight read active, prefill and verify previously called the per-projection `_Fp8BlockScaleWeightReadLinear` view modules, and now issue the fused `qkvg_proj` GEMM plus g-slice reuse for the output gate — even with `KIMI_K3_KDA_GLUE_FP8=0`. The 128-block quantization argument in `quantize_weight`'s docstring says the results should be identical, but the two new GPU tests only exercise the BF16 `_qkvg_proj_weight` branch. Consider extending one parity test to run after `_convert_kda_projections_to_fp8_weight_read` (random weights suffice), or record in the PR that an FP8 config was validated manually.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py [hunk 3]
```diff
@@ -1027,23 +1029,37 @@ def finalize_decode_weights(self) -> None:
...
             )
-            fused = torch.cat([m.weight.data for m in mods], dim=0).contiguous()
-            off = 0
-            for m in mods:
-                n = m.weight.shape[0]
-                m.weight.data = fused[off : off + n]
-                off += n
+            qkvg_weight = self._merge_projection_weights(qkvg_modules)
+            bfa_weight = self._merge_projection_weights(
+                (mixer.f_a_proj, mixer.b_proj), pad_rows_to=8
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_kimi_linear.py:1047]: Worth a one-line comment on why `pad_rows_to=8`: nothing at the definition or either call site says what the alignment buys (GEMV kernel selection for the [f_a|b] tail, presumably). The next reader can't tell whether 8 is a hard kernel requirement or a measured tuning choice, which matters if head_dim/num_heads change on a future config.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py [hunk 4]
```diff
@@ -108,6 +108,7 @@
 )
 
 _KDA_INDEXED_STATE_POOL_ENABLED = os.environ.get("TLLM_KDA_ENABLE_INDEXED_STATE_POOL", "1") == "1"
+_KDA_BFA_MULTISTREAM_MAX_ROWS = 128
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_kimi_linear.py:116]: Document the basis for 128 here — is it measured (overlap stops paying once the wide qkvg GEMM occupies most SMs) or a heuristic? A one-line comment saying which, and on what hardware it was tuned, keeps someone from cargo-culting or "fixing" it later. The verify path also compares total rows (`num_decodes * num_steps`) against it, so at MTP num_steps=3 the cutoff is ~42 requests — worth confirming that's the intended tradeoff rather than a per-request bound.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py (hunk 2 above)
Comment (jiaganc) [reply tensorrt_llm/_torch/models/modeling_kimi_linear.py:1262]: I don't want to add fp8 test now, since these fp8 knob will likely be removed after modelopt release quant checkpoint. We will have a new fp8 path then.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py (hunk 3 above)
Comment (jiaganc) [reply tensorrt_llm/_torch/models/modeling_kimi_linear.py:1047]: Added clarification in a192d810dc. Eight BF16 outputs occupy 16 bytes, so padding keeps each fused-output row aligned for vectorized `f_b` consumption. It is a performance tuning choice, not a hard kernel requirement.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py (hunk 1 above)
Comment (jiaganc) [reply tensorrt_llm/_torch/models/modeling_kimi_linear.py:1709]: Thanks for reviewing. We have plan to investigate and remove these `contiguous()`s afterwards, so keep it here now.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py (hunk 4 above)
Comment (jiaganc) [reply tensorrt_llm/_torch/models/modeling_kimi_linear.py:116]: Documented in f98990f389. The 128-row cutoff is now identified as a heuristic ported from SGLang’s Blackwell path, with a permalink to the source and a note that it still needs TensorRT-LLM-specific benchmarking and tuning. SGLang validated the overlap on 2x4 GB300 TP8 at batch sizes 1 and 64; 128 itself was not established as a measured crossover. Verify intentionally uses `B * num_steps` because those flattened token rows are the projection GEMMs’ M dimension, so with three MTP steps the effective cutoff is indeed about 42 requests.
